zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/http/ElasticsearchHttpSpanConsumer.java

```diff
@@ -93,9 +112,22 @@
 
   void putServiceSpans(Map<String, Set<Pair<String>>> indexToServiceSpans, String index, Span s) {
     Set<Pair<String>> serviceSpans = indexToServiceSpans.get(index);
-    if (serviceSpans == null) indexToServiceSpans.put(index, serviceSpans = new LinkedHashSet<>());
-    for (String serviceName : s.serviceNames()) {
-      serviceSpans.add(Pair.create(serviceName, s.name));
+    synchronized(indexToServiceSpansCache) {
```

Comment:
you may end up needing to tune this later.. maybe for now put in some unit tests, then later some contended benchmarks